go/ngql/types.go

```diff
@@ -207,12 +273,68 @@ func structToGQLObject(nomsType *types.Type, tm *typeMap) *graphql.Object {
 	})
 }
 
+func listAndSetToGraphQLInputObject(nomsType *types.Type, tm *typeMap) graphql.Input {
+	nomsValueType := nomsType.Desc.(types.CompoundDesc).ElemTypes[0]
+	if nomsValueType.Equals(types.MakeUnionType()) {
+		return nil
+	}
+	elemType := NomsTypeToGraphQLInputType(nomsValueType, tm)
+	if elemType == nil {
+		return nil
+	}
+	return graphql.NewList(graphql.NewNonNull(elemType))
+}
+
+func mapToGraphQLInputObject(nomsType *types.Type, tm *typeMap) graphql.Input {
+	nomsKeyType := nomsType.Desc.(types.CompoundDesc).ElemTypes[0]
+	nomsValueType := nomsType.Desc.(types.CompoundDesc).ElemTypes[1]
+
+	if nomsKeyType.Equals(types.MakeUnionType()) || nomsValueType.Equals(types.MakeUnionType()) {
+		return nil
+	}
+
+	keyType := NomsTypeToGraphQLInputType(nomsKeyType, tm)
+	valueType := NomsTypeToGraphQLInputType(nomsValueType, tm)
+	if keyType == nil || valueType == nil {
+		return nil
+	}
+
+	entryType := mapEntryToGraphQLInputObject(keyType, valueType, nomsKeyType, nomsValueType, tm)
+	return graphql.NewList(entryType)
+}
+
+func structToGQLInputObject(nomsType *types.Type, tm *typeMap) graphql.Input {
+	// GraphQL input types cannot contain cycles.
+	unresolved := types.ToUnresolvedType(nomsType)
+	if unresolved.HasUnresolvedCycle() {
+		return nil
```

Comment:
Not sure where to report these. At the moment if we end up with nil for the input type we do not add the arguments (key, keys, through). Unknown errors are errors in GrpahQL (we still have that validation on) but it might be confusing.

Failing when run into these is not an option because we run into these when we automatically create the schema based on the data in the dataset.

One option might be to return an error from NomsTypeToGraphQLInputType but ignore that in the auto schema generation... I'll give it a try..